docker/elasticsearch-setup/create-indices.sh

```diff
@@ -5,71 +5,96 @@ set -e
 : ${USE_AWS_ELASTICSEARCH:=false}
 
+# protocol: http or https?
 if [[ $ELASTICSEARCH_USE_SSL == true ]]; then
     ELASTICSEARCH_PROTOCOL=https
 else
     ELASTICSEARCH_PROTOCOL=http
 fi
 
-if [[ ! -z $ELASTICSEARCH_USERNAME ]] && [[ -z $ELASTICSEARCH_AUTH_HEADER ]]; then
-  AUTH_TOKEN=$(echo -ne "$ELASTICSEARCH_USERNAME:$ELASTICSEARCH_PASSWORD" | base64 --wrap 0)
-  ELASTICSEARCH_AUTH_HEADER="Authorization:Basic $AUTH_TOKEN"
-fi
+# Elasticsearch URL to be suffixed with a resource address
+ELASTICSEARCH_URL="$ELASTICSEARCH_PROTOCOL://$ELASTICSEARCH_HOST:$ELASTICSEARCH_PORT"
 
-# Add default header if needed
+# set auth header if none is given
 if [[ -z $ELASTICSEARCH_AUTH_HEADER ]]; then
-  ELASTICSEARCH_AUTH_HEADER="Accept: */*"
-fi
-
-function create_datahub_usage_event_datastream() {
-  if [[ -z "$INDEX_PREFIX" ]]; then
-    PREFIX=''
+  if [[ ! -z $ELASTICSEARCH_USERNAME ]]; then
+    # no auth header given, but username is defined -> use it to create the auth header
+    AUTH_TOKEN=$(echo -ne "$ELASTICSEARCH_USERNAME:$ELASTICSEARCH_PASSWORD" | base64 --wrap 0)
+    ELASTICSEARCH_AUTH_HEADER="Authorization:Basic $AUTH_TOKEN"
   else
-    PREFIX="${INDEX_PREFIX}_"
+    # no auth header or username given -> use default auth header
+    ELASTICSEARCH_AUTH_HEADER="Accept: */*"
   fi
+fi
+
+# index prefix used throughout the script
+if [[ -z "$INDEX_PREFIX" ]]; then
+  PREFIX=''
+else
+  PREFIX="${INDEX_PREFIX}_"
+fi
+
+# path where index definitions are stored
+INDEX_DEFINITIONS_ROOT=/index/usage-event
+
+
+# check Elasticsearch for given index/resource (first argument)
+# if it doesn't exist (http code 404), use the given file (second argument) to create it
+function create_if_not_exists {
+  RESOURCE_ADDRESS="$1"
+  RESOURCE_DEFINITION_NAME="$2"
+
+  # query ES to see if the resource already exists
+  RESOURCE_STATUS=$(curl -o /dev/null -s -w "%{http_code}\n" --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_URL/$RESOURCE_ADDRESS")
+
+  if [ $RESOURCE_STATUS -eq 200 ]; then
+    # resource already exists -> nothing to do
+    echo -e "\n>>> $RESOURCE_ADDRESS already exists ✓"
+
+  elif [ $RESOURCE_STATUS -eq 404 ]; then
+    # resource doesn't exist -> need to create it
+    echo -e "\n>>> creating $RESOURCE_ADDRESS ..."
+    # use the given path as definition, but first replace all occurences of PREFIX with the actual prefix
+    TMP_SOURCE_PATH="/tmp/$RESOURCE_DEFINITION_NAME"
+    sed -e "s/PREFIX/$PREFIX/g" "$INDEX_DEFINITIONS_ROOT/$RESOURCE_DEFINITION_NAME" | tee -a "$TMP_SOURCE_PATH"
+    curl -s -XPUT --header "$ELASTICSEARCH_AUTH_HEADER" "$ELASTICSEARCH_URL/$RESOURCE_ADDRESS" -H 'Content-Type: application/json' --data "@$TMP_SOURCE_PATH"
+
+  elif [ $RESOURCE_STATUS -eq 401 ] || [ $RESOURCE_STATUS -eq 405 ]; then
+    echo -e "\n>>> failed to GET $RESOURCE_ADDRESS ($RESOURCE_STATUS) !"
+    echo "... make sure you have correct USE_AWS_ELASTICSEARCH env value set (current=$USE_AWS_ELASTICSEARCH)"
```

Comment:
Why AWS specific?

Reply:
Not having `USE_AWS_ELASTICSEARCH` set to `true` is the common issue several people have met, which then causes Analytics to go wrong. I observed the 401/405 errors to be linked with this particular misconfiguration and figured it might be helpful for people to be aware of the obvious fix. (See the log in **Testing / Case 2** section of the PR description.)

However I can see this looks too specific for a general-purpose function. Let me add a condition checking whether `$ELASTICSEARCH_URL` contains some aws-specific substring and display the message only then.